cli/command/image/build.go

```diff
@@ -368,7 +439,7 @@ type resolvedTag struct {
 // "FROM <image>" instructions to a digest reference. `translator` is a
 // function that takes a repository name and tag reference and returns a
 // trusted digest reference.
-func rewriteDockerfileFrom(ctx context.Context, dockerfile io.Reader, translator translatorFunc) (newDockerfile []byte, resolvedTags []*resolvedTag, err error) {
+func rewriteDockerfileFrom(ctx context.Context, dockerfile io.Reader, translator translatorFunc, observer observerFunc) (newDockerfile []byte, resolvedTags []*resolvedTag, err error) {
```

Comment:
It seems like there is more code in the two branches (translator vs observer) than is actually being shared by the two use cases.

I think this could be improved by extracting a function like this:

```go

type visitor func(line string) error
func walkDockerfileLines(dockerfile io.Reader, visit visitor) error {
   ...
}

func getFromReference(line string) (ref string, ok bool) {
   ...
}
```

`walkDockerfileLines()` could be called with a closure in `rewriteDockerfileFrom()` to create the `[]byte`, and could be called with an observer to get the references.